pkg/egressgateway/manager.go

```diff
@@ -265,23 +355,63 @@ func (manager *Manager) OnUpdateEndpoint(endpoint *k8sTypes.CiliumEndpoint) {
 		return
 	}
 
+	if _, ok := manager.epDataStore[epData.id]; ok {
+		logger.Debug("Updated CiliumEndpoint")
+	} else {
+		logger.Debug("Added CiliumEndpoint")
+	}
+
 	manager.epDataStore[epData.id] = epData
 
 	manager.reconcile(eventUpdateEndpoint)
 }
 
+// OnUpdateEndpoint is the event handler for endpoint additions and updates.
+func (manager *Manager) OnUpdateEndpoint(endpoint *k8sTypes.CiliumEndpoint) {
+	id := types.NamespacedName{
+		Name:      endpoint.GetName(),
+		Namespace: endpoint.GetNamespace(),
+	}
+
+	// remove the ID from the workqueue in case an older revision of the
+	// same CiliumEndpoint was already present
+	manager.endpointEventRetryQueue.Forget(id)
+	manager.endpointEventRetryQueue.Done(id)
```

Comment:
It is a bit strange seeing these operations being performed outside of the reconcile loop, but I cannot spot anything that could go wrong with them. Still, I wonder if it could make sense to just add the entry to the pending events and push the ID in the queue, leaving the reconcile loop to trigger `addEndpoint`.

Reply:
I think this should stay here as it has nothing to do with the reconciliation logic. More in general, the way the manager is structured is the following:
* we have a bunch of event handlers called by the k8s watchers (for policies, endpoints and nodes events) which simply parse the events and update the copies of these resources stored in the manager
* and then we have the reconciliation logic which goes through these objects and ensure the state of the node is in sync with these resources

because of this I think parsing an endpoint event (and retrying because we failed to resolve identity -> labels) should not be part of the reconciliation logic

Reply:
Yeah, I agree, and I think we might be referring to different things when mentioning "reconciliation logic". In particular, I was meaning the `retryEndpointIdentityLabelsResolution` loop, and not `manager.reconcile()` (I agree that the latter is independent and should not handle the endpoint parsing for symmetry with the other handlers). 

Still, my comment was mostly a nit, and I'm also fine with the current version.

Reply:
Ah, I see what you mean now :+1: 
I changed the `On{Update,Delete}Endpoint` handlers to only add and remove items from the workqueue/pending events map, and let the retry loop do the rest (so call `addEnpoint` and deal with re-adding back items or marking them as done)